[mono-move] Add module loader with versioned cache slots (#19506)
aptos-labs/aptos-core: third_party/move/mono-move/loader/README.md

# MonoMove Loader Design

## 1. Context

The MonoMove runtime uses a long-living module cache that satisfies the following requirements.

> - Stores executable IR and lowered code to avoid repeated deserializion, verification and translaction.

Reviewer: nit: 
```suggestion
- Stores executable IR and lowered code to avoid repeated deserialization, verification and translation.
```

- Supports upgrades: 1) within same block, 2) between speculative block trees (with Zaptos, blocks with published code may not be committed immediately).
  Because of speculation and concurrent upgrades, newer or speculative code can co-exist with older versions at the same time.

When a single transaction executes, it may "load" modules into its local read-set.
The read-set is used to cache a consistent local view of the cache for transaction.
This is needed to avoid cases where some transaction upgrades module `A`, and other transaction ends up seeing both old and new versions at the same time.

Loading a module can either be a hit in long-living module cache or a miss.
On a miss, loader fetches the module from storage, deserializes it, verifies, translates to execution IR, and adds it to the long-living cache.
To cover for the worst case, every module load has to be charged gas.

> Importantny, the charging must be **deterministic** irrespective of the cache state that may be different on different validators due to speculation.

Reviewer: nit:
```suggestion
Importantly, the charging must be **deterministic** irrespective of the cache state that may be different on different validators due to speculation.
```

Transaction may need to load modules for multiple reasons.

1. When executing a cross-module call.
2. When lowering non-generic function IR to micro-ops (struct layout information such as size, field offsets is needed).
3. When monomorphizing generic function IR to micro-ops (same as for non-generics, layout information is needed).

Note that (2) and (3) also cover cases like 1) deserializing data on global storage access, 2) walking VM values based on layouts (serialization, GC).

## 2. Challenges

> 1. **Loading costs are determinisitc.**

Reviewer: nit:

```suggestion
1. **Loading costs are deterministic.**
```

   Two executions of the same transaction must charge the same cost regardless of what the long-living cache stores.
   As a result, transactions need to enforce cache hit has the same gas charging semantics as the cache miss.

2. **Loading costs are charged on first access.**

>    Multiple accesses to same module `A` should be charged on first access only.

Reviewer 2: ```suggestion
   Multiple accesses to same module `A` should be charged for it's loading costs on the first access only.
```

   While repeated accessed can be overcharged by a small factor, it can easily become too restrictive for contracts.
   If there is a loop that calls into module `A`, gas must not be charged for `A` on every iteration.
   Similarly, if `A` is a hot library called from many places within a transaction, charging for all call sites would be prohibitively expensive for call-heavy Move programs.

3. **Layout costs depend on multiple modules.**
   In order to lower a function, specializer needs to know layouts of struct types.
   Layout information is aggregated over multiple modules, so its cost depends on all of them.
   This problem is tightly coupled with *charge on first access*.
   For example, if transaction has loaded and charged for modules `A`, `B`, and `C`, it cannot charge aggregated cost for layout that uses `A`, `B`, and `D`.
   Instead, only extra `D` has to be charged.
   As a result, semantics of gas charging may impact how efficient loading is.

4. **Modules are upgradable.**

>    On upgrade, modules may change their code and size.

Reviewer 2: on upgrade, types can also change (structs be added, enums be extended), which also may need to be interned when loaded.

For example, consider a transaction that has `A`, `B`, and `C` charged.
Suppose that is accesses layout information from cache that depends on `A`, `B`, and `D`.
Because of metering, it needs to only charge for non-visited subset of the modules.
But, computing this subset is effectively a pull-based validation.

### Loading policies

When loading a module, it might not be possible to lower its non-generic functions to micro-ops because layouts from other modules are needed.
For example, consider the following collections of Move modules.
If module `a` is loaded, `f1` cannot be lowered until layout of `A1` is known.
That in turn requires modules `b` and `c`.

```move
// Example 1

module 0x23::a {
   struct A1 { x: 0x23::b::B1, y: u64 }
   struct A2 { x: u64, y: u64 }
   public fun f1(x: &A1): u64 { x.y }
   public fun f2(x: &A2): u64 { x.y }
}

module 0x23::b {
   struct B1 { x: 0x23::c::C, y: u64 }
   struct B2 { x: 0x23::d::D, y: u64 }
   public fun g(x: &B2): u64 { x.y }

}

module 0x23::c {
   struct C { x: u64 }
}

module 0x23::d {
   struct D { x: bool }
}
```

> In this design, we propose three loadig policies.

Reviewer 2: ```suggestion
In this design, we propose three loading policies.
```

│   └── B2 → B3 → B4        ← parent chain has reconfig, empty blocks
└── B1' (no reconfig)
    └── B2'                  ← parent chain has no reconfig, executes normally
```

### Mandatory Sets

All three loading policies can be unified under the same concept: **mandatory sets (MS)**.
Mandatory set is a set of modules of fixed size that has to be loaded on the first module access or during the first function call.
In Example 1, under LL policy `MS(a) = {a}`, `MS(f1) = {b, c}`, and `MS(f2) = {}`.
Under EL policy, `MS(a) = {a, b, c}`, `MS(f1) = {}`, and `MS(f2) = {}`.
Under PL policy, mandatory set is all modules in the package.

In order to make metering deterministic:

1. When module `m` is loaded, gas is charged for all modules in `MS(m)` not in current read-set.
2. When function `f` is called (after its module has been loaded), gas is charged for all modules in `MS(f)` not in current read-set.
3. All charged modules are recorded in a per-transaction read-set.

Here, (2) also handles metering for generic calls because type argument layouts are in mandatory set of the caller.

The goal of mandatory sets is to enforce determinism under a particular loading policy.

> Depending on the policy, module loading may be more or less expensive, making function call resultion less or more expensive.

Reviewer 2: ```suggestion
Depending on the policy, module loading may be more or less expensive, making function call resolution less or more expensive.
```

The best policy is workload dependent.
Interpreter can also use inline caches to avoid repeated set intersection checks, if they are expensive.

> The important proerty of mandatory sets is that they cannot change size because of upgrade.

Reviewer: nit:
```suggestion
The important property of mandatory sets is that they cannot change size because of upgrade.
```

> This makes them particularly atractive because they can be cached per each module instance.

Reviewer 2: ```suggestion
This makes them particularly attractive because they can be cached per each module instance.
```

> For example, if module `a` is cached in long-living cache, there is no need to re-traverse its depndencies in some complex way to simulate EL loading and metering.

Reviewer: nit:
```suggestion
For example, if module `a` is cached in long-living cache, there is no need to re-traverse its dependencies in some complex way to simulate EL loading and metering.
```

```rust
pub struct Slot<T> {
    /// Committed baseline value (storage).
    base: AtomicPtr<T>,        
    /// If true, need to resolve via overlay + pending.
    stale: AtomicBool,
    // Writes from all executed blocks in the speculative tree, compressed to at most 1 version per block.
    overlay: Mutex<SmallVec<[(BlockId, *const T); 2]>>,
    // Current block writes in-flight.
    pending: Mutex<SmallVec<[(TxnIdx, *const T); 1]>>, 
}
```

Mandatory sets are implemented as a slice of pointers to `Slot` instances in the cache.
Slots are stable and never claimed by garbage collection unless the cache is flushed.

### Handling Cache Misses for Loading Policies

When loader encounters a cache miss, it needs to load the mandatory set of the target module.
These loads can also be cache misses and have to be inserted into long-living cache.
In order to enforce correctness, loader splits "loading", "linking" and "insertion" into their own phases.

1. A set of cache misses is obtained (corresponding to the mandatory set).
   The modules can still be in compiled, file-format representation.
2. Modules are translated to execution IR.
   Non-generic functions are lowered when possible.

> 3. Modules are ordered in reversed topological order.

Reviewer 2: Reverse topological order of what? Dependencies? And why is this ordering critical for correctness under concurrent loads?

Author: Dependencies, fixed that.It is important if we link:

Say you have modules `A` and `B` where `A` links to `B`, both in same package. Both cache misses. Now, you want calls from `A` to `B` to use direct pointers, so you patch pointers in `A` without inserting `B` yet. What happens next:
- Other thread inserts `B` in the cache (now slot is taken)
- Your thread inserts `B` but because of the race it is a loser, so no-op, `B` is freed.
- Your thread does succeed in inserting `A` but now your pointers are dangling.

Topological order on dependencies is just. way to enforce that you inserted a module you link agains so it is the same across all threads, and is guaranteed to be alive. This is just one way to do it. Conceptually, we can also have a package lock, and insert all modules, then release the lock. But then cache has to be package aware, as reading some module `A` will block on lock taken, and currently we have 0 info on what package module belongs to. We can revisit the implementation as we integrate package metadata, but these are roughly 2 options.
